Python/Product/Analysis/LanguageServer/Server.cs

```diff
@@ -613,49 +603,44 @@ private sealed class ReloadModulesQueueItem : IAnalyzable {
 
-            var diags = _analyzer.GetDiagnostics(pythonProjectEntry);
-            var severity = _settings?.diagnostics?.unresolvedImports ?? DiagnosticSeverity.Warning;
-            if (parse != null && severity.Show()) {
-                var walker = new ImportStatementWalker(parse.Tree, pythonProjectEntry, _analyzer, severity);
-                parse.Tree.Walk(walker);
-                diags = diags.Concat(walker.Diagnostics).ToArray();
-            }
-
-            if (!diags.Any()) {
-                return;
-            }
-
-            if (pythonProjectEntry is IDocument doc) {
-                var reported = _openFiles.GetDocument(doc.DocumentUri).LastReportedDiagnostics;
-                lock (reported) {
-                    if (reported.TryGetValue(0, out var lastVersion)) {
-                        diags = diags.Concat(lastVersion.Diagnostics).ToArray();
-                    }
-                }
-            }
-
-            PublishDiagnostics(new PublishDiagnosticsEventArgs {
-                diagnostics = diags,
-                uri = pythonProjectEntry.DocumentUri,
-                _version = version
-            });
+            _editorFiles.GetDocument(pythonProjectEntry.DocumentUri).UpdateAnalysisDiagnostics(pythonProjectEntry);
         }
 
         private PythonAst GetParseTree(IPythonProjectEntry entry, Uri documentUri, CancellationToken token, out BufferVersion bufferVersion) {
             PythonAst tree = null;
             bufferVersion = null;
-            var parse = entry.WaitForCurrentParse(_clientCaps?.python?.completionsTimeout ?? Timeout.Infinite, token);
+            var parse = entry.WaitForCurrentParse(Timeout.Infinite, token);
             if (parse != null) {
                 tree = parse.Tree;
                 if (parse.Cookie is VersionCookie vc) {
-                    if (vc.Versions.TryGetValue(_projectFiles.GetPart(documentUri), out bufferVersion)) {
+                    if (vc.Versions.TryGetValue(ProjectFiles.GetPart(documentUri), out bufferVersion)) {
                         tree = bufferVersion.Ast ?? tree;
                     }
                 }
             }
             return tree;
         }
 
+        private bool HandleConfigurationChanges(LanguageServerSettings newSettings) {
+            var oldSettings = Settings;
+            Settings = newSettings;
+
+            if (oldSettings == null) {
+                return true;
+            }
+
+            if (newSettings.analysis.openFilesOnly != oldSettings.analysis.openFilesOnly) {
+                _editorFiles.UpdateDiagnostics();
+            }
+
+            if (!newSettings.analysis.errors.SequenceEqual(oldSettings.analysis.errors) ||
+                !newSettings.analysis.warnings.SequenceEqual(oldSettings.analysis.warnings) ||
+                !newSettings.analysis.information.SequenceEqual(oldSettings.analysis.information) ||
+                !newSettings.analysis.disabled.SequenceEqual(oldSettings.analysis.disabled)) {
```

Comment:
We should use set equality here, not sequence equality. (`errors.Except(oldErrors).Any()`)

Reply:
What is old is longer than the new one? Would it make new.Except(old).Any empty?

Reply:
`old.Intersect(new).Any()` I guess

Reply:
Yeah, you're right. Turn them into sets and do SetEquals() I guess